spec/services/json_validator_spec.rb

```diff
@@ -0,0 +1,517 @@
+require "spec_helper"
+
+RSpec.describe JsonValidator do
+  include FormResponseHelper
+
+  let(:data) { valid_data }
+  let(:schema_data) { File.read(Rails.root.join("config/schemas/form_response.json")) }
+  let(:schema) { JSON.parse schema_data }
+  subject { described_class.new(schema, data) }
+
+  describe "#invalid?" do
+    it "returns false if data is valid" do
+      expect(subject.invalid?).to be(false)
+    end
+
+    context "with invalid data" do
+      let(:data) { valid_data.except(:live_in_england) }
+
+      it "returns true" do
+        expect(subject.invalid?).to be(true)
+      end
+    end
+  end
+
+  describe "#errors" do
+    let(:error) { subject.errors.first }
+
+    it "returns no error if the data is valid" do
+      expect(subject.errors).to be_empty
+    end
+
+    context "when live_in_england missing" do
+      let(:data) { valid_data.except(:live_in_england) }
+
+      it "returns a missing keys error" do
+        expect(error["missing_keys"]).to contain_exactly("live_in_england")
+        expect(error["data"]).to be_nil
+      end
+    end
+
+    context "when live_in_england in valid" do
+      let(:data) { valid_data.merge(live_in_england: "invalid") }
```

Comment:
```suggestion
      let(:invalid_data) { valid_data.merge(live_in_england: "invalid") }
```

I think "data" is an overloaded term as it's also something that is returned in `errors`

Reply:
I've renamed it `json_data`. I was tempted to call it `json` but as it's actually a Hash, added the `_data` suffix.